Code/Framework/AzCore/AzCore/DOM/DomPatch.cpp

```diff
@@ -0,0 +1,959 @@
+        if (!pathLookup.IsSuccess())
+        {
+            return AZ::Failure(pathLookup.TakeError());
+        }
+
+        if (!Utils::DeepCompareIsEqual(rootElement[m_domPath], GetValue()))
+        {
+            return AZ::Failure<AZStd::string>("Test failed, values don't match");
+        }
+
+        return AZ::Success();
+    }
+
+    namespace PatchApplicationStrategy
+    {
+        void HaltOnFailure(PatchApplicationState& state)
+        {
+            if (!state.m_outcome.IsSuccess())
+            {
+                state.m_shouldContinue = false;
+            }
+        }
+
+        void IgnoreFailureAndContinue([[maybe_unused]] PatchApplicationState& state)
+        {
+        }
+    } // namespace PatchApplicationStrategy
+
+    Patch::Patch(AZStd::initializer_list<PatchOperation> init)
+        : m_operations(init)
+    {
+    }
+
+    bool Patch::operator==(const Patch& rhs) const
+    {
+        if (m_operations.size() != rhs.m_operations.size())
+        {
+            return false;
+        }
+
+        for (size_t i = 0; i < m_operations.size(); ++i)
+        {
+            if (m_operations[i] != rhs.m_operations[i])
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    bool Patch::operator!=(const Patch& rhs) const
+    {
+        return !operator==(rhs);
+    }
+
+    const Patch::OperationsContainer& Patch::GetOperations() const
+    {
+        return m_operations;
+    }
+
+    void Patch::PushBack(PatchOperation op)
+    {
+        m_operations.push_back(AZStd::move(op));
+    }
+
+    void Patch::PushFront(PatchOperation op)
+    {
+        m_operations.insert(m_operations.begin(), AZStd::move(op));
+    }
+
+    void Patch::Pop()
+    {
+        m_operations.pop_back();
+    }
+
+    void Patch::Clear()
+    {
+        m_operations.clear();
+    }
+
+    const PatchOperation& Patch::At(size_t index) const
+    {
+        return m_operations[index];
+    }
+
+    size_t Patch::Size() const
+    {
+        return m_operations.size();
+    }
+
+    PatchOperation& Patch::operator[](size_t index)
+    {
+        return m_operations[index];
+    }
+
+    const PatchOperation& Patch::operator[](size_t index) const
+    {
+        return m_operations[index];
+    }
+
+    Patch::OperationsContainer::iterator Patch::begin()
+    {
+        return m_operations.begin();
+    }
+
+    Patch::OperationsContainer::iterator Patch::end()
+    {
+        return m_operations.end();
+    }
+
+    Patch::OperationsContainer::const_iterator Patch::begin() const
+    {
+        return m_operations.cbegin();
+    }
+
+    Patch::OperationsContainer::const_iterator Patch::end() const
+    {
+        return m_operations.cend();
+    }
+
+    Patch::OperationsContainer::const_iterator Patch::cbegin() const
+    {
+        return m_operations.cbegin();
+    }
+
+    Patch::OperationsContainer::const_iterator Patch::cend() const
+    {
+        return m_operations.cend();
+    }
```

Comment:
The need to specify `Patch` class as part of the return type can be alleviated by using a trailing return type

```suggestion
    auto Patch::begin() -> OperationsContainer::iterator
    {
        return m_operations.begin();
    }

    auto Patch::end() -> OperationsContainer::iterator
    {
        return m_operations.end();
    }

    auto Patch::begin() const -> OperationsContainer::const_iterator
    {
        return m_operations.cbegin();
    }

    auto Patch::end() const -> OperationsContainer::const_iterator
    {
        return m_operations.cend();
    }

    auto Patch::cbegin() const -> OperationsContainer::const_iterator
    {
        return m_operations.cbegin();
    }

    auto Patch::cend() const -> OperationsContainer::const_iterator
    {
        return m_operations.cend();
    }
```

Reply:
I hadn't thought about using trailing return type for that 👍 will use in the future as well